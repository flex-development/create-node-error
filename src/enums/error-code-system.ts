/**
 * @file Enums - SystemErrorCode
 * @module errnode/enums/SystemErrorCode
 */

/**
 * System error codes.
 *
 * @see https://github.com/nodejs/node/blob/v19.3.0/deps/uv/include/uv.h#L72-L154
 *
 * @enum {Uppercase<string>}
 */
enum SystemErrorCode {
  E2BIG = 'E2BIG',
  EACCES = 'EACCES',
  EADDRINUSE = 'EADDRINUSE',
  EADDRNOTAVAIL = 'EADDRNOTAVAIL',
  EAFNOSUPPORT = 'EAFNOSUPPORT',
  EAGAIN = 'EAGAIN',
  EAI_ADDRFAMILY = 'EAI_ADDRFAMILY',
  EAI_AGAIN = 'EAI_AGAIN',
  EAI_BADFLAGS = 'EAI_BADFLAGS',
  EAI_BADHINTS = 'EAI_BADHINTS',
  EAI_CANCELED = 'EAI_CANCELED',
  EAI_FAIL = 'EAI_FAIL',
  EAI_FAMILY = 'EAI_FAMILY',
  EAI_MEMORY = 'EAI_MEMORY',
  EAI_NODATA = 'EAI_NODATA',
  EAI_NONAME = 'EAI_NONAME',
  EAI_OVERFLOW = 'EAI_OVERFLOW',
  EAI_PROTOCOL = 'EAI_PROTOCOL',
  EAI_SERVICE = 'EAI_SERVICE',
  EAI_SOCKTYPE = 'EAI_SOCKTYPE',
  EALREADY = 'EALREADY',
  EBADF = 'EBADF',
  EBADMSG = 'EBADMSG',
  EBUSY = 'EBUSY',
  ECANCELED = 'ECANCELED',
  ECHARSET = 'ECHARSET',
  ECHILD = 'ECHILD',
  ECONNABORTED = 'ECONNABORTED',
  ECONNREFUSED = 'ECONNREFUSED',
  ECONNRESET = 'ECONNRESET',
  EDEADLK = 'EDEADLK',
  EDESTADDRREQ = 'EDESTADDRREQ',
  EDOM = 'EDOM',
  EDQUOT = 'EDQUOT',
  EEXIST = 'EEXIST',
  EFAULT = 'EFAULT',
  EFBIG = 'EFBIG',
  EFTYPE = 'EFTYPE',
  EHOSTDOWN = 'EHOSTDOWN',
  EHOSTUNREACH = 'EHOSTUNREACH',
  EIDRM = 'EIDRM',
  EILSEQ = 'EILSEQ',
  EINPROGRESS = 'EINPROGRESS',
  EINTR = 'EINTR',
  EINVAL = 'EINVAL',
  EIO = 'EIO',
  EISCONN = 'EISCONN',
  EISDIR = 'EISDIR',
  ELOOP = 'ELOOP',
  EMFILE = 'EMFILE',
  EMLINK = 'EMLINK',
  EMSGSIZE = 'EMSGSIZE',
  EMULTIHOP = 'EMULTIHOP',
  ENAMETOOLONG = 'ENAMETOOLONG',
  ENETDOWN = 'ENETDOWN',
  ENETRESET = 'ENETRESET',
  ENETUNREACH = 'ENETUNREACH',
  ENFILE = 'ENFILE',
  ENOBUFS = 'ENOBUFS',
  ENODATA = 'ENODATA',
  ENODEV = 'ENODEV',
  ENOENT = 'ENOENT',
  ENOEXEC = 'ENOEXEC',
  ENOLCK = 'ENOLCK',
  ENOLINK = 'ENOLINK',
  ENOMEM = 'ENOMEM',
  ENOMSG = 'ENOMSG',
  ENONET = 'ENONET',
  ENOPROTOOPT = 'ENOPROTOOPT',
  ENOSPC = 'ENOSPC',
  ENOSR = 'ENOSR',
  ENOSTR = 'ENOSTR',
  ENOSYS = 'ENOSYS',
  ENOTCONN = 'ENOTCONN',
  ENOTDIR = 'ENOTDIR',
  ENOTEMPTY = 'ENOTEMPTY',
  ENOTSOCK = 'ENOTSOCK',
  ENOTSUP = 'ENOTSUP',
  ENOTTY = 'ENOTTY',
  ENXIO = 'ENXIO',
  EOF = 'EOF',
  EOPNOTSUPP = 'EOPNOTSUPP',
  EOVERFLOW = 'EOVERFLOW',
  EPERM = 'EPERM',
  EPIPE = 'EPIPE',
  EPROTO = 'EPROTO',
  EPROTONOSUPPORT = 'EPROTONOSUPPORT',
  EPROTOTYPE = 'EPROTOTYPE',
  ERANGE = 'ERANGE',
  EREMOTEIO = 'EREMOTEIO',
  EROFS = 'EROFS',
  ESHUTDOWN = 'ESHUTDOWN',
  ESOCKTNOSUPPORT = 'ESOCKTNOSUPPORT',
  ESPIPE = 'ESPIPE',
  ESRCH = 'ESRCH',
  ESTALE = 'ESTALE',
  ETIME = 'ETIME',
  ETIMEDOUT = 'ETIMEDOUT',
  ETXTBSY = 'ETXTBSY',
  EUNATCH = 'EUNATCH',
  EWOULDBLOCK = 'EWOULDBLOCK',
  EXDEV = 'EXDEV',
  UNKNOWN = 'UNKNOWN'
}

export default SystemErrorCode
