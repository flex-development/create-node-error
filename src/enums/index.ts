/**
 * @file Entry Point - Enums
 * @module errnode/enums
 */

export { default as codes } from './codes'
export { default as syscodes } from './syscodes'
