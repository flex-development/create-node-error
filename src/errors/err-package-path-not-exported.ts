/**
 * @file Errors - ERR_PACKAGE_PATH_NOT_EXPORTED
 * @module errnode/errors/ERR_PACKAGE_PATH_NOT_EXPORTED
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L1632-L1638
 */

import E from '#e'
import { codes } from '#src/enums'
import type { NodeError, NodeErrorConstructor } from '#src/interfaces'
import { DOT } from '@flex-development/tutils'
import { ok } from 'devlop'

/**
 * `ERR_PACKAGE_PATH_NOT_EXPORTED` schema.
 *
 * @see {@linkcode NodeError}
 * @see https://nodejs.org/api/errors.html#err_package_path_not_exported
 *
 * @extends {NodeError<codes.ERR_PACKAGE_PATH_NOT_EXPORTED>}
 */
interface ErrPackagePathNotExported
  extends NodeError<codes.ERR_PACKAGE_PATH_NOT_EXPORTED> {}

/**
 * `ERR_PACKAGE_PATH_NOT_EXPORTED` message arguments.
 */
type Args = [dir: string, subpath: string, base?: string | null | undefined]

/**
 * `ERR_PACKAGE_PATH_NOT_EXPORTED` constructor.
 *
 * @see {@linkcode Args}
 * @see {@linkcode ErrPackagePathNotExported}
 * @see {@linkcode NodeErrorConstructor}
 *
 * @extends {NodeErrorConstructor<ErrPackagePathNotExported,Args>}
 */
interface ErrPackagePathNotExportedConstructor
  extends NodeErrorConstructor<ErrPackagePathNotExported, Args> {
  /**
   * Create a new `ERR_PACKAGE_PATH_NOT_EXPORTED` error.
   *
   * @see {@linkcode ErrPackagePathNotExported}
   *
   * @param {string} dir
   *  Id of package directory
   * @param {string} subpath
   *  Requested subpath
   * @param {string | null | undefined} [base]
   *  Parent module path
   * @return {ErrPackagePathNotExported}
   */
  new (
    dir: string,
    subpath: string,
    base?: string | null | undefined
  ): ErrPackagePathNotExported
}

/**
 * `ERR_PACKAGE_PATH_NOT_EXPORTED` model.
 *
 * Thrown when a `package.json` [`"exports"`][1] field does not export the
 * requested subpath.
 *
 * [1]: https://nodejs.org/api/packages.html#exports
 *
 * @see {@linkcode ErrPackagePathNotExportedConstructor}
 *
 * @type {ErrPackagePathNotExportedConstructor}
 *
 * @class
 */
const ERR_PACKAGE_PATH_NOT_EXPORTED: ErrPackagePathNotExportedConstructor = E(
  codes.ERR_PACKAGE_PATH_NOT_EXPORTED,
  Error,
  /**
   * @param {string} dir
   *  Id of package directory
   * @param {string} subpath
   *  Requested subpath
   * @param {string | null | undefined} [base]
   *  Parent module path
   * @return {string} Error message
   */
  function message(
    dir: string,
    subpath: string,
    base: string | null | undefined = null
  ): string {
    /**
     * Error message.
     *
     * @var {string} message
     */
    let message: string = subpath === DOT
      ? 'No \'exports\' main defined in'
      : `Package subpath '${subpath}' is not defined by 'exports' in`

    ok(dir.endsWith('/'), 'expected `dir` to end with trailing slash ("/")')
    message += ` ${dir}package.json`
    if (base) message += ` imported from ${base}`

    return message
  }
)

export {
  ERR_PACKAGE_PATH_NOT_EXPORTED as default,
  type ErrPackagePathNotExported,
  type Args as ErrPackagePathNotExportedArgs,
  type ErrPackagePathNotExportedConstructor
}
