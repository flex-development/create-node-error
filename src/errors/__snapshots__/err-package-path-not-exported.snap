// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/ERR_PACKAGE_PATH_NOT_EXPORTED > should build message from parameters (0) 1`] = `No 'exports' main defined in \${process.cwd()}/node_modules/@flex-development/pathe/package.json imported from \${process.cwd()}/src/errors/__tests__/err-package-path-not-exported.spec.ts`;

exports[`unit:errors/ERR_PACKAGE_PATH_NOT_EXPORTED > should build message from parameters (1) 1`] = `Package subpath './package.json' is not defined by 'exports' in \${process.cwd()}/node_modules/@flex-development/pathe/package.json imported from \${process.cwd()}/src/errors/__tests__/err-package-path-not-exported.spec.ts`;
